pkg/apis/authentication/validation/credentialsbinding.go

```diff
@@ -0,0 +1,113 @@
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMeta(&binding.ObjectMeta, true, gardencorevalidation.ValidateName, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateCredentials(binding.CredentialsRef, field.NewPath("credentialsRef"))...)
+	allErrs = append(allErrs, ValidateCredentialsBindingProvider(binding.Provider, field.NewPath("provider"))...)
+	for i, quota := range binding.Quotas {
+		allErrs = append(allErrs, validateObjectReferenceOptionalNamespace(quota, field.NewPath("quotas").Index(i))...)
+	}
+
+	return allErrs
+}
+
+// ValidateCredentialsBindingUpdate validates a CredentialsBinding object before an update.
+func ValidateCredentialsBindingUpdate(newBinding, oldBinding *authentication.CredentialsBinding) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&newBinding.ObjectMeta, &oldBinding.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newBinding.Quotas, oldBinding.Quotas, field.NewPath("quotas"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newBinding.Provider, oldBinding.Provider, field.NewPath("provider"))...)
+
+	allErrs = append(allErrs, ValidateCredentialsBinding(newBinding)...)
+
+	return allErrs
+}
+
+// ValidateCredentialsBindingProvider validates a CredentialsBindingProvider object.
+func ValidateCredentialsBindingProvider(provider authentication.CredentialsBindingProvider, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(provider.Type) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("type"), "must specify a provider type"))
+	}
+
+	if len(strings.Split(provider.Type, ",")) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("type"), provider.Type, "multiple providers specified"))
+	}
+
+	return allErrs
+}
+
+func validateObjectReferenceOptionalNamespace(ref corev1.ObjectReference, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(ref.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "must provide a name"))
+	}
+
+	return allErrs
+}
+
+func validateCredentials(ref corev1.ObjectReference, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(ref.APIVersion) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("apiVersion"), "must provide an apiVersion"))
+	}
+
+	if len(ref.Kind) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("kind"), "must provide a kind"))
+	}
+
+	if len(ref.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), "must provide a name"))
+	}
+
+	for _, err := range validation.IsDNS1123Subdomain(ref.Name) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), ref.Name, err))
+	}
+
+	if len(ref.Namespace) == 0 {
```

Comment:
It seems that namespace is required, but then in the documentation we mark it as optional https://github.com/gardener/gardener/pull/9626/files#diff-1def7650a3a5fc5a5c82768eb49af4ee0763dc14d7c57a0e52aea0dbada39cd4R14 🤔

Reply:
It is defaulted to the CredentialsBinding namespace if unset, https://github.com/gardener/gardener/pull/9626/files#diff-be308c2a17f43103bfacf7fa296b79c3fed327e3431c1911f262a0fea885438fR10. 

So, from client side point of view it is optional, but from API server point of view it is required. Maybe the check can be skipped as defaulting takes care for it, but it also does not harm. Anyway, if you insist - I will remove it, what do you think?

Reply:
Thanks for clarifying, I missed the defaulting part. No need to remove it, it makes sense to have the check.